Support User: Bypass localStorage when active

jordwest opened the pull request:
This change keeps all localStorage data from a support user session isolated from the main user session.

To achieve this, `localStorage` is bypassed and replaced with an in-memory store - only while support user is active. This in combination with #3879 will mean we can avoid having to refresh all open sessions, closing #3819.

-------

### How to test

1. Set `localStorage.setItem( 'debug', 'calypso:support-user' )`
2. Log in to support user (ref p4TIVU-2OW-p2)
3. When login is successful and Calypso reboots, messages in the console should indicate `localStorage` bypass.
4. Check that the actual stored localStorage values (using the browser dev tools) relate to the original user, not the support user. Specifically, you can check that `wpcom_user` shows the display name of the original user.
5. Using the browser console, run the following command:
```js
JSON.parse( localStorage.getItem( 'wpcom_user' ) );
```
The console should show the support user's data.

Tested in:
 - [x] Chrome 48
 - [x] Firefox 44
 - [x] Safari 9
 - [x] IE/Edge
 - [x] With `support-user` feature flag disabled
 - [x] In local `production` environment

gwwar commented on client/lib/user/support-user-interop.js:
```diff
@@ -99,6 +99,55 @@ export const boot = () => {
+		// they are safe to share across user sessions.
+		const allowedKeys = [ STORAGE_KEY, 'debug' ];
+
+		Storage.prototype.setItem = ( key, value ) => {
```
Since we're overriding the Storage interface, we should be careful and override each interface: https://html.spec.whatwg.org/multipage/webstorage.html#the-storage-interface
```
interface Storage {
  readonly attribute unsigned long length;
  DOMString? key(unsigned long index);
  getter DOMString? getItem(DOMString key);
  setter void setItem(DOMString key, DOMString value);
  deleter void removeItem(DOMString key);
  void clear();
};
```

It looks like we're missing `key(n)` and the length attribute

gwwar commented:
:+1: verified that this works around localStorage, and behaves normally when we switch back. I highly recommend adding tests.

In the long run, a cleaner solution would be to move over store/localStorage usages to localforage and create a special in-memory driver so we no longer need to override the Storage prototype. 

cc @retrofox We still need to verify sync-handler behavior before closing #3819. It may look pretty similar to #3879

jordwest commented:
@gwwar: I've moved the localStorage override into its own module and added unit tests similar to those in the localForage override PR (#3977).

Once there is nothing in Calypso using localStorage directly, we should be able to revert the changes in this PR.

gwwar commented on client/lib/support/support-user/localstorage-bypass.js:
```diff
@@ -0,0 +1,91 @@
+};
+
+export default function( allowedKeys ) {
+	if ( window && window.localStorage && window.Storage && window.Storage.prototype ) {
```
Maybe leave a comment here on why we're overriding the Storage interface.

gwwar commented on client/lib/support/support-user/test/localstorage-bypass.js:
```diff
@@ -0,0 +1,134 @@
+	length,
+} from '../localstorage-bypass';
+
+describe( 'localstorage-bypass', () => {
```
To enable this test in the single test runner, we'll need to whitelist this by adding:
```json
		"support": {
			"support-user": {
				"test": [ "localstorage-bypass" ]
			}
		},
```

to `client/tests.json`.

Also see: https://github.com/Automattic/wp-calypso/blob/master/test/README.md

gwwar commented on client/lib/support/support-user/test/localstorage-bypass.js:
```diff
@@ -0,0 +1,134 @@
+ * External dependencies
+ */
+import { expect } from 'chai';
+import { spy } from 'sinon';
```
The usage here is fine, but we also have a number of test helpers now in our single test runner, including: https://github.com/Automattic/wp-calypso/tree/master/test/test/helpers/use-sinon

gwwar commented:
:+1: good to :shipit: after adding the test to the whitelist.
